libs/vkd3d/shaders/cs_workgraph_distribute_payload_offsets.comp

```diff
@@ -0,0 +1,265 @@
+
+layout(buffer_reference, buffer_reference_align = 8, std430) readonly buffer NodePayloadOffsetCount
+{
+	uvec2 data[];
+};
+
+layout(buffer_reference, buffer_reference_align = 4, std430) writeonly buffer UnrolledOffsets
+{
+	uint data[];
+};
+
+#include "cs_workgraph_data_structures.h"
+
+// Abuse aliasing rules to make sure that we can get scalar loads while doing
+// atomics to part of the buffer :v
+layout(buffer_reference, buffer_reference_align = 16, std430) buffer IndirectCommandsBufferAtomic
+{
+	layout(offset = 16) IndirectCommands indirect_commands_atomic[];
+};
+
+layout(buffer_reference, buffer_reference_align = 16, std430) restrict readonly buffer IndirectCommandsBufferRO
+{
+	layout(offset = 12) uint total_fused_elements;
+	IndirectCommands indirect_commands_read[];
+};
+
+// For patching in sharing count.
+layout(buffer_reference, buffer_reference_align = 4, std430) buffer Payload32
+{
+	uint data[];
+};
+
+layout(buffer_reference, buffer_reference_align = 4, std430) buffer Payload16
+{
+	uint16_t data[];
+};
+
+struct NodeMeta
+{
+	uint packed_control;
+	uint payload_stride_grid_offset_or_count;
+};
+
+layout(buffer_reference, buffer_reference_align = 8, std430) restrict readonly buffer NodeTypeMeta
+{
+	NodeMeta data[];
+};
+
+layout(push_constant, std430) uniform Registers
+{
+	NodePayloadOffsetCount packed_offset_counts;
+	UnrolledOffsets unrolled_offsets;
+	IndirectCommandsBufferRO commands;
+	Payload32 payload;
+	NodeTypeMeta meta;
+} registers;
+
+void process_node_index(uint node_index, uint payload_offset, uint count)
+{
+	NodeMeta node_meta = registers.meta.data[node_index];
+	int group_components = int(bitfieldExtract(node_meta.packed_control, 0, 8));
+	bool group_components_u32 = bitfieldExtract(node_meta.packed_control, 8, 8) != 0;
+	bool rw_group_tracking = bitfieldExtract(node_meta.packed_control, 16, 8) != 0;
+	bool group_compact_broadcast = bitfieldExtract(node_meta.packed_control, 24, 8) != 0;
+	int grid_offset_or_count = bitfieldExtract(int(node_meta.payload_stride_grid_offset_or_count), 16, 16);
+	uint payload_stride = bitfieldExtract(node_meta.payload_stride_grid_offset_or_count, 0, 16);
+
+	bool should_compact_broadcast = group_compact_broadcast;
+	if (should_compact_broadcast)
+		should_compact_broadcast = registers.commands.indirect_commands_read[node_index].primary_execute.y == 0u;
+
+	// If we're not going to compact, we can allocate atomically once.
+	uint output_offset;
+	if (!should_compact_broadcast)
+	{
+		uint total_scan = subgroupAdd(count);
+		if (subgroupElect())
+			output_offset = atomicAdd(IndirectCommandsBufferAtomic(registers.commands).indirect_commands_atomic[node_index].linear_offset_atomic, total_scan);
+		output_offset = subgroupBroadcastFirst(output_offset);
+	}
+
+	uvec4 ballot = subgroupBallot(count != 0u);
+
+	while (any(notEqual(ballot, uvec4(0))))
+	{
+		uint lane = subgroupBallotFindLSB(ballot);
+
+		// TODO: Is there a more elegant way that is just as fast and fully portable?
```

Comment:
only thing that would come to mind is rewriting the loop as something like

```glsl
bool thread_mask = count != 0;

while (subgroupAny(thread_mask)) {
  uint lane = subgroupBallotFindLSB(subgroupBallot(thread_mask));
  thread_mask = thread_mask && gl_SubgroupInvocationID > lane;
  ...
}
```

Reply:
I think that comparison would require vector ops.